File: rust/src/cmd/reopen.rs
use std::path::Path;

use clap::Args;

use crate::git;
use crate::thread::{self, Thread};
use crate::workspace;

#[derive(Args)]
pub struct ReopenArgs {
    /// Thread ID or name reference
    id: String,

    /// Status to reopen to
    #[arg(long, default_value = "active")]
    status: String,

    /// Commit after reopening
    #[arg(long)]
    commit: bool,

    /// Commit message
    #[arg(short = 'm', long)]
    m: Option<String>,
}

pub fn run(args: ReopenArgs, ws: &Path) -> Result<(), String> {
    if !thread::is_valid_status(&args.status) {
        return Err(format!(
            "Invalid status '{}'. Must be one of: idea, planning, active, blocked, paused, resolved, superseded, deferred",
            args.status
        ));
    }

    let file = workspace::find_by_ref(ws, &args.id)?;

    let mut t = Thread::parse(&file)?;

    let old_status = t.status().to_string();

    // Update status
    t.set_frontmatter_field("status", &args.status)?;

    // Add log entry
    t.content = thread::insert_log_entry(&t.content, "Reopened.");

    t.write()?;

    println!("Reopened: {} → {} ({})", old_status, args.status, file.display());

    if args.commit {
        let msg = args.m.unwrap_or_else(|| {
            git::generate_commit_message(ws, &[file.to_string_lossy().to_string()])
        });
        git::auto_commit(ws, &file, &msg)?;
    } else {
        println!(
            "Note: Thread {} has uncommitted changes. Use 'threads commit {}' when ready.",
            args.id, args.id
        );
    }

    Ok(())
}

File: go/internal/cmd/reopen.go
package cmd

import (
	"fmt"

	"github.com/spf13/cobra"

	"git.zib.de/cspiegel/threads/internal/git"
	"git.zib.de/cspiegel/threads/internal/thread"
	"git.zib.de/cspiegel/threads/internal/workspace"
)

var (
	reopenStatus string
	reopenCommit bool
	reopenMsg    string
)

var reopenCmd = &cobra.Command{
	Use:               "reopen <id>",
	Short:             "Reopen resolved thread",
	Args:              cobra.ExactArgs(1),
	ValidArgsFunction: completeThreadIDs,
	RunE:              runReopen,
}

func init() {
	reopenCmd.Flags().StringVar(&reopenStatus, "status", "active", "Status to reopen to")
	reopenCmd.Flags().BoolVar(&reopenCommit, "commit", false, "Commit after reopening")
	reopenCmd.Flags().StringVarP(&reopenMsg, "m", "m", "", "Commit message")
}

func runReopen(cmd *cobra.Command, args []string) error {
	ws := getWorkspace()
	ref := args[0]

	file, err := workspace.FindByRef(ws, ref)
	if err != nil {
		return err
	}

	t, err := thread.Parse(file)
	if err != nil {
		return err
	}

	oldStatus := t.Status()

	// Update status
	if err := t.SetFrontmatterField("status", reopenStatus); err != nil {
		return err
	}

	// Add log entry
	t.Content = thread.InsertLogEntry(t.Content, "Reopened.")

	if err := t.Write(); err != nil {
		return err
	}

	fmt.Printf("Reopened: %s → %s (%s)\n", oldStatus, reopenStatus, file)

	if reopenCommit {
		msg := reopenMsg
		if msg == "" {
			msg = git.GenerateCommitMessage(ws, []string{file})
		}
		if err := git.AutoCommit(ws, file, msg); err != nil {
			return err
		}
	} else {
		fmt.Printf("Note: Thread %s has uncommitted changes. Use 'threads commit %s' when ready.\n", ref, ref)
	}

	return nil
}
